### PR TITLE
install: warn if WantedBy targets don't exist

### PR DESCRIPTION
**Pull request by dtardon**
Currently, if [Install] section contains WantedBy=target that doesn't exist,
systemd creates the symlinks anyway. That is just user-unfriendly.
Let's be nice and warn about installing non-existent targets.

Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1835351.

Replaces: #15834

### CODE REVIEW

File: src/shared/install.c [hunk 1]
```diff
@@ -3461,6 +3471,7 @@ static const char* const unit_file_change_type_table[_UNIT_FILE_CHANGE_TYPE_MAX]
         [UNIT_FILE_UNLINK]      = "unlink",
         [UNIT_FILE_IS_MASKED]   = "masked",
         [UNIT_FILE_IS_DANGLING] = "dangling",
+        [UNIT_FILE_IS_UNKNOWN_TARGET]     = "unknown target",
```
Comment (bluca) [inline src/shared/install.c]: re-align the table please

Comment (dtardon) [reply src/shared/install.c]: Done. But I see a lot of similar tables in other places that are not aligned...

**Comment by dtardon** [comment]
The test failures appear unrelated...

File: src/test/test-install-root.c [hunk 2]
```diff
@@ -1244,6 +1244,12 @@ int main(int argc, char *argv[]) {
         p = strjoina(root, "/usr/lib/systemd/system-preset/");
         assert_se(mkdir_p(p, 0755) >= 0);
 
+        p = strjoina(root, "/usr/lib/systemd/system/multi-user.target");
+        assert_se(write_string_file(p, "", WRITE_STRING_FILE_CREATE) >= 0);
```
Comment (keszybz) [inline src/test/test-install-root.c]: Hmm, but an empty file is a mask, i.e. the same as symlinking to /dev/null. Shouldn't there be at least a comment so the file is not empty?

File: src/shared/install.c [hunk 3]
```diff
@@ -3457,10 +3467,11 @@ static const char* const unit_file_state_table[_UNIT_FILE_STATE_MAX] = {
...
 static const char* const unit_file_change_type_table[_UNIT_FILE_CHANGE_TYPE_MAX] = {
-        [UNIT_FILE_SYMLINK]     = "symlink",
-        [UNIT_FILE_UNLINK]      = "unlink",
-        [UNIT_FILE_IS_MASKED]   = "masked",
-        [UNIT_FILE_IS_DANGLING] = "dangling",
+        [UNIT_FILE_SYMLINK]             = "symlink",
+        [UNIT_FILE_UNLINK]              = "unlink",
+        [UNIT_FILE_IS_MASKED]           = "masked",
+        [UNIT_FILE_IS_DANGLING]         = "dangling",
+        [UNIT_FILE_IS_UNKNOWN_TARGET]   = "unknown target",
```
Comment (keszybz) [inline src/shared/install.c]: Now there's an extra empty column…

**Review by keszybz — COMMENTED** [review]
Looks good, some minor comments.

File: src/shared/install.c (hunk 3 above)
Comment (dtardon) [reply src/shared/install.c]: Fixed.

File: src/test/test-install-root.c (hunk 2 above)
Comment (dtardon) [reply src/test/test-install-root.c]: Fixed.

**Comment by keszybz** [comment]
LGTM.

File: src/shared/install.h [hunk 4]
```diff
@@ -32,6 +32,7 @@ enum {
         UNIT_FILE_UNLINK,
         UNIT_FILE_IS_MASKED,
         UNIT_FILE_IS_DANGLING,
+        UNIT_FILE_IS_UNKNOWN_TARGET,
```
Comment (poettering) [inline src/shared/install.h]: i don't like the name of this. the specified unit can be of any type, not just target units

Comment (poettering) [reply src/shared/install.h]: maybe: "vacant"? in the sense of "squatting a vacant lot"

**Comment by poettering** [comment]
lgtm, but i'd like to find a better name for the returned code. "target" is misleading or wrong, it can be any unit type. maybe @keszybz has an idea how to call this? i like "vacant", but it's not ideal.

**Comment by keszybz** [comment]
"vacant" would be a property of the destination itself, but here we're talking about the destination not being present at all. So maybe call this `UNIT_FILE_DESTINATION_NOT_PRESENT` ?

File: src/shared/install.h (hunk 4 above)
Comment (dtardon) [reply src/shared/install.h]: Well, it's supposed to mean dependency target, not a unit type...

**Comment by dtardon** [comment]
> "vacant" would be a property of the destination itself, but here we're talking about the destination not being present at all. So maybe call this `UNIT_FILE_DESTINATION_NOT_PRESENT` ?

I've used this.

**Comment by keszybz** [comment]
mkosi / ci (arch): Failed to dissect image '/home/runner/work/systemd/systemd/image.raw': Connection timed out

**Pull request merged**